src/compiler/checker.ts

```diff
@@ -18117,8 +18117,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     removeFromEach(typeSet, TypeFlags.Null);
                     result = getUnionType([getIntersectionType(typeSet, flags), nullType], UnionReduction.Literal, aliasSymbol, aliasTypeArguments);
                 }
-                else if (typeSet.length >= 4) {
-                    // When we have four or more constituents, some of which are unions, we employ a "divide and conquer" strategy
+                else if (typeSet.length >= 3 && types.length > 2) {
+                    // When we have three or more constituents, more than two inputs (to head off infinite reexpansion), some of which are unions, we employ a "divide and conquer" strategy

```

Comment:
can you give an example of the infinite re-expansion?
Does it occur because `typeSet.slice(middle)` might recur right back to this case, repeatedly when `typeSet.slice(middle).length === 1`?

Reply:
> Does it occur because typeSet.slice(middle) might recur right back to this case, repeatedly when typeSet.slice(middle).length === 1?

Yep. If the input is one type, but that one type expands into multiple typeSet members when inspected (because it is itself an intersection containing a union), we can end up back here as we ping-pong between trying to distribute intersections to within unions while also maximally breaking down these intersection powerset to pairwise combinations. Just bailing on this optimization if we only got a single type as input is a very low-overhead way to short-circuit that loop, since our splitting will (eventually) split down to, at the worst, a pair of types on one side plus a single type on the other.